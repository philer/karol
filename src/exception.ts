/**
 * Errors for our simulated programming language and runtime.
 * These do not inherit from JS's own Error as they do not need
 * to reveal details of the interpreter/runtime internals.
 */
export class Exception {
  message: string
  data?: Record<string, any>
  constructor(message: string, data?: Record<string, any>) {
    this.message = message
    this.data = data
  }
}
