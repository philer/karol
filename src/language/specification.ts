import * as config from "../config"
import {getLocaleData} from "../localization"
import type {Builtins} from "../simulation/world"
import {identity} from "../util"


export const TokenType = {
  IDENTIFIER: "IDENTIFIER",
  INTEGER: "INTEGER",
  NOT: "NOT",
  IF: "IF",
  THEN: "THEN",
  ELSE: "ELSE",
  WHILE: "WHILE",
  DO: "DO",
  REPEAT: "REPEAT",
  TIMES: "TIMES",
  PROGRAM: "PROGRAM",
  ROUTINE: "ROUTINE",
  LPAREN: "LPAREN",
  RPAREN: "RPAREN",
  LBRACKET: "LBRACKET",
  RBRACKET: "RBRACKET",
  // LBRACE: "LBRACE",
  // RBRACE: "RBRACE",
  LESS: "LESS",
  GREATER: "GREATER",
  EQUALS: "EQUALS",
  ASTERISC: "ASTERISC",
  SLASH: "SLASH",
  HYPHENMINUS: "HYPHENMINUS",
  PLUS: "PLUS",
  DOT: "DOT",
  COMMA: "COMMA",
  COLON: "COLON",
  SEMI: "SEMI",
  SINGLEQUOTE: "SINGLEQUOTE",
  DOUBLEQUOTE: "DOUBLEQUOTE",
  WHITESPACE: "WHITESPACE",
  COMMENT: "COMMENT",
  EOF: "EOF",
} as const

export type TokenType = keyof typeof TokenType

export const symbolToTokenType: Readonly<Record<string, TokenType>> = {
  "(": TokenType.LPAREN,
  ")": TokenType.RPAREN,
  "[": TokenType.LBRACKET,
  "]": TokenType.RBRACKET,
  // "{": TokenType.LBRACE,  // comments
  // "}": TokenType.RBRACE,  // comments
  "<": TokenType.LESS,
  ">": TokenType.GREATER,
  "=": TokenType.EQUALS,
  "*": TokenType.ASTERISC,
  "/": TokenType.SLASH,
  "-": TokenType.HYPHENMINUS,
  "+": TokenType.PLUS,
  ".": TokenType.DOT,
  ",": TokenType.COMMA,
  ":": TokenType.COLON,
  ";": TokenType.SEMI,
  "'": TokenType.SINGLEQUOTE,
  '"': TokenType.DOUBLEQUOTE,
}

export type LanguageSpecification = Readonly<{
  keywords: Readonly<Record<string, TokenType>>
  builtins: Readonly<Record<string, keyof Builtins>>
  normalizeIdentifier: (identifier: string) => string
  normalizeKeyword: (keyword: string) => string
}>

const toLowerCase = (str: string) => str.toLowerCase()

export const defaultSpec: LanguageSpecification = {
  keywords: {
    if: "IF",
    then: "THEN",
    else: "ELSE",
    while: "WHILE",
    do: "DO",
    not: "NOT",
    repeat: "REPEAT",
    times: "TIMES",
    program: "PROGRAM",
    routine: "ROUTINE",
  },
  builtins: {
    isEdge: "isLookingAtEdge",
    notIsEdge: "isNotLookingAtEdge",
    step: "step",
    stepBackwards: "stepBackwards",
    isLookingNorth: "isLookingNorth",
    isLookingEast: "isLookingEast",
    isLookingSouth: "isLookingSouth",
    isLookingWest: "isLookingWest",
    turnLeft: "turnLeft",
    turnRight: "turnRight",
    isBlock: "isLookingAtBlock",
    notIsBlock: "isNotLookingAtBlock",
    placeBlock: "placeBlock",
    takeBlock: "takeBlock",
    isOnMark: "isOnMark",
    notIsOnMark: "isNotOnMark",
    placeMark: "placeMark",
    takeMark: "takeMark",
  },
  normalizeKeyword: identity,
  normalizeIdentifier: identity,
}

/**
 * Load a customized programming language specification
 */
export async function load(): Promise<LanguageSpecification> {
  const {code} = await config.get()
  const locales = await Promise.all(code.locales.map(getLocaleData))
  const normalizeKeyword = code.caseSensitiveKeywords ? identity : toLowerCase
  const normalizeIdentifier = code.caseSensitiveIdentifiers ? identity : toLowerCase

  const keywords = {} as Record<string, TokenType>
  const builtins = {} as Record<string, keyof Builtins>
  for (const {language} of locales) {
    for (const [tt, translations] of Object.entries(language.keywords)) {
      for (const trans of translations) {
        keywords[trans] = tt as TokenType
        keywords[normalizeKeyword(trans)] = tt as TokenType
      }
    }
    for (const [cmd, translations] of Object.entries(language.builtins)) {
      for (const trans of translations) {
        builtins[trans] = cmd as keyof Builtins
        builtins[normalizeIdentifier(trans)] = cmd as keyof Builtins
      }
    }
  }

  return {keywords, builtins, normalizeKeyword, normalizeIdentifier}
}
