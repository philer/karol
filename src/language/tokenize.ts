import {Exception} from "../exception"
import type {LanguageSpecification, TokenType} from "./specification"
// eslint-disable-next-line no-duplicate-imports
import {symbolToTokenType, TokenType as tt} from "./specification"

export interface Token {
  type: TokenType
  isBuiltin?: boolean
  value: string
  line: number
  column: number
}

/**
 * Iterable lexer
 */
export function* tokenize(
  text: string,
  {keywords, builtins, normalizeKeyword, normalizeIdentifier}: LanguageSpecification,
  yieldWhitespace = false,
  yieldComments = false,
): Generator<Token, Token> {
  const length = text.length

  const reSpaces = /\s+/y
  const reComment = /\{.*?\}/sy
  const reSinglelineComment = /\/\/.*/y
  const reInteger = /\d+/y

  // https://stackoverflow.com/questions/30225552/regex-for-diacritics/44586972
  // https://stackoverflow.com/questions/30798522/regular-expression-not-working-for-at-least-one-european-character
  const reWord = /[A-zÀ-ÖØ-öø-įĴ-őŔ-žǍ-ǰǴ-ǵǸ-țȞ-ȟȤ-ȳɃɆ-ɏḀ-ẞƀ-ƓƗ-ƚƝ-ơƤ-ƥƫ-ưƲ-ƶẠ-ỿ][A-zÀ-ÖØ-öø-įĴ-őŔ-žǍ-ǰǴ-ǵǸ-țȞ-ȟȤ-ȳɃɆ-ɏḀ-ẞƀ-ƓƗ-ƚƝ-ơƤ-ƥƫ-ưƲ-ƶẠ-ỿ0-9]*/y

  let position = 0
  let line = 1
  let column = 1

  let value = ""
  let match: RegExpExecArray | null
  let lines: string[]

  while (position < length) {

    // word (identifier / keyword)
    reWord.lastIndex = position
    if (match = reWord.exec(text)) {
      value = match[0]
      yield {
        type: keywords[normalizeKeyword(value)] || tt.IDENTIFIER,
        isBuiltin: builtins.hasOwnProperty(normalizeIdentifier(value)),
        value, line, column,
      }
      column += value.length
      position += value.length
      continue
    }

    // whitespace
    reSpaces.lastIndex = position
    if (match = reSpaces.exec(text)) {
      value = match[0]
      if (yieldWhitespace) {
        yield {type: tt.WHITESPACE, value, line, column}
      }
      lines = value.split("\n")
      if (lines.length > 1) {
        line += lines.length - 1
        column = lines[lines.length - 1].length + 1
      } else {
        column += value.length
      }
      position += value.length
      continue
    }

    // multi-line comment
    reComment.lastIndex = position
    if (match = reComment.exec(text)) {
      value = match[0]
      if (yieldComments) {
        yield {type: tt.COMMENT, value, line, column}
      }
      lines = value.split("\n")
      if (lines.length > 1) {
        line += lines.length - 1
        column = lines[lines.length - 1].length + 1
      } else {
        column += value.length
      }
      position += value.length
      continue
    }

    // single-line comment
    reSinglelineComment.lastIndex = position
    if (match = reSinglelineComment.exec(text)) {
      value = match[0]
      if (yieldComments) {
        yield {type: tt.COMMENT, value, line, column}
      }
      column += value.length
      position += value.length
      continue
    }

    // special character (must be checked after // comments)
    value = text[position]
    if (symbolToTokenType.hasOwnProperty(value)) {
      yield {type: symbolToTokenType[value], value, line, column}
      ++column
      ++position
      continue
    }

    // integer
    reInteger.lastIndex = position
    if (match = reInteger.exec(text)) {
      value = match[0]
      yield {type: tt.INTEGER, value, line, column}
      column += value.length
      position += value.length
      continue
    }

    // found nothing useful
    throw new Exception("error.parser.token_read", {
      line,
      column,
      remainingText: text.slice(position),
    })
  }
  return {type: tt.EOF, value: "", line, column}
}
