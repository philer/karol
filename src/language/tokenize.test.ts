import {mapObject} from "../util"
import {Exception} from "../exception"
import {tokenize} from "./tokenize"
import {LanguageSpecification, defaultSpec} from "./specification"

describe("tokenizer", () => {

  test("empty program", () => {
    expect([...tokenize("", defaultSpec)])
      .toEqual([])
    expect([...tokenize("", defaultSpec, true, true)])
      .toEqual([])
  })

  test("identifiers", () => {
    expect([...tokenize("ascii", defaultSpec)]).toEqual([{
      type: "IDENTIFIER", line: 1, column: 1,
      value: "ascii",
      isBuiltin: false,
    }])
    const utf8 = "ÁéèôæçäöüÄÖÜß"
    expect([...tokenize(utf8, defaultSpec)]).toEqual([{
      type: "IDENTIFIER", line: 1, column: 1,
      value: utf8,
      isBuiltin: false,
    }])
  })

  test("normalization", () => {
    type Kws = LanguageSpecification["keywords"]
    type Bltns = LanguageSpecification["builtins"]
    const spec: LanguageSpecification = {
      keywords: {
        ...defaultSpec.keywords,
        ...mapObject<Kws, Kws>((trans, tt) => [trans.toLowerCase(), tt], defaultSpec.keywords),
      },
      builtins: {
        ...defaultSpec.builtins,
        ...mapObject<Bltns, Bltns>((trans, cmd) => [trans.toLowerCase(), cmd], defaultSpec.builtins),
      },
      normalizeKeyword: str => str.toLowerCase(),
      normalizeIdentifier: str => str.toLowerCase(),
    }
    expect([...tokenize("while", spec)]).toEqual([{
      type: "WHILE", line: 1, column: 1,
      value: "while",
      isBuiltin: false,
    }])
    expect([...tokenize("wHiLe", spec)]).toEqual([{
      type: "WHILE", line: 1, column: 1,
      value: "wHiLe",
      isBuiltin: false,
    }])
    expect([...tokenize("isEdge", spec)]).toEqual([{
      type: "IDENTIFIER", line: 1, column: 1,
      value: "isEdge",
      isBuiltin: true,
    }])
    expect([...tokenize("isedge", spec)]).toEqual([{
      type: "IDENTIFIER", line: 1, column: 1,
      value: "isedge",
      isBuiltin: true,
    }])
    expect([...tokenize("custom", spec)]).toEqual([{
      type: "IDENTIFIER", line: 1, column: 1,
      value: "custom",
      isBuiltin: false,
    }])
  })

  test("integers", () => {
    expect([...tokenize("0", defaultSpec)])
      .toEqual([{type: "INTEGER", value: "0", line: 1, column: 1}])
    expect([...tokenize("1234567890", defaultSpec)])
      .toEqual([{type: "INTEGER", value: "1234567890", line: 1, column: 1}])
    expect([...tokenize("00123", defaultSpec)])
      .toEqual([{type: "INTEGER", value: "00123", line: 1, column: 1}])
  })

  test("program 'wall'", () => {
    expect([...tokenize(code, defaultSpec, true, true)]).toMatchSnapshot()
    expect([...tokenize(code, defaultSpec)]).toMatchSnapshot()
  })

  test("returns EOF", () => {
    expect(tokenize("", defaultSpec).next())
      .toEqual({done: true, value: {type: "EOF", value: "", line: 1, column: 1}})
  })

  test("unknown characters", () => {
    expect(() => [...tokenize("·", defaultSpec)]).toThrow(Exception)
  })
})

const code = `{ Build a wall around the world }

routine wall()
    while not isEdge() do
        placeBlock(2)
        step()
    *while
*routine

// Stairs
placeBlock()
step()
placeBlock(2)
step()
turnRight() turnRight()
placeBlock()
turnLeft() turnLeft()

wall()
turnLeft()
wall()
turnLeft()
wall()
turnLeft()
wall()
turnLeft()
`
