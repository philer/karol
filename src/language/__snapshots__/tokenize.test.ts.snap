// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer program 'wall' 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "type": "COMMENT",
    "value": "{ Build a wall around the world }",
  },
  Object {
    "column": 34,
    "line": 1,
    "type": "WHITESPACE",
    "value": "

",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 3,
    "type": "ROUTINE",
    "value": "routine",
  },
  Object {
    "column": 8,
    "line": 3,
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "column": 9,
    "isBuiltin": false,
    "line": 3,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 13,
    "line": 3,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 14,
    "line": 3,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 15,
    "line": 3,
    "type": "WHITESPACE",
    "value": "
    ",
  },
  Object {
    "column": 5,
    "isBuiltin": false,
    "line": 4,
    "type": "WHILE",
    "value": "while",
  },
  Object {
    "column": 10,
    "line": 4,
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "column": 11,
    "isBuiltin": false,
    "line": 4,
    "type": "NOT",
    "value": "not",
  },
  Object {
    "column": 14,
    "line": 4,
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "column": 15,
    "isBuiltin": true,
    "line": 4,
    "type": "IDENTIFIER",
    "value": "isEdge",
  },
  Object {
    "column": 21,
    "line": 4,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 22,
    "line": 4,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 23,
    "line": 4,
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "column": 24,
    "isBuiltin": false,
    "line": 4,
    "type": "DO",
    "value": "do",
  },
  Object {
    "column": 26,
    "line": 4,
    "type": "WHITESPACE",
    "value": "
        ",
  },
  Object {
    "column": 9,
    "isBuiltin": true,
    "line": 5,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 19,
    "line": 5,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 20,
    "line": 5,
    "type": "INTEGER",
    "value": "2",
  },
  Object {
    "column": 21,
    "line": 5,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 22,
    "line": 5,
    "type": "WHITESPACE",
    "value": "
        ",
  },
  Object {
    "column": 9,
    "isBuiltin": true,
    "line": 6,
    "type": "IDENTIFIER",
    "value": "step",
  },
  Object {
    "column": 13,
    "line": 6,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 14,
    "line": 6,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 15,
    "line": 6,
    "type": "WHITESPACE",
    "value": "
    ",
  },
  Object {
    "column": 5,
    "line": 7,
    "type": "ASTERISC",
    "value": "*",
  },
  Object {
    "column": 6,
    "isBuiltin": false,
    "line": 7,
    "type": "WHILE",
    "value": "while",
  },
  Object {
    "column": 11,
    "line": 7,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "line": 8,
    "type": "ASTERISC",
    "value": "*",
  },
  Object {
    "column": 2,
    "isBuiltin": false,
    "line": 8,
    "type": "ROUTINE",
    "value": "routine",
  },
  Object {
    "column": 9,
    "line": 8,
    "type": "WHITESPACE",
    "value": "

",
  },
  Object {
    "column": 1,
    "line": 10,
    "type": "COMMENT",
    "value": "// Stairs",
  },
  Object {
    "column": 10,
    "line": 10,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 11,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 11,
    "line": 11,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 12,
    "line": 11,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 13,
    "line": 11,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 12,
    "type": "IDENTIFIER",
    "value": "step",
  },
  Object {
    "column": 5,
    "line": 12,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 12,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 7,
    "line": 12,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 13,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 11,
    "line": 13,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 12,
    "line": 13,
    "type": "INTEGER",
    "value": "2",
  },
  Object {
    "column": 13,
    "line": 13,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 14,
    "line": 13,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 14,
    "type": "IDENTIFIER",
    "value": "step",
  },
  Object {
    "column": 5,
    "line": 14,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 14,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 7,
    "line": 14,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 15,
    "type": "IDENTIFIER",
    "value": "turnRight",
  },
  Object {
    "column": 10,
    "line": 15,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 11,
    "line": 15,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 12,
    "line": 15,
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "column": 13,
    "isBuiltin": true,
    "line": 15,
    "type": "IDENTIFIER",
    "value": "turnRight",
  },
  Object {
    "column": 22,
    "line": 15,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 23,
    "line": 15,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 24,
    "line": 15,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 16,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 11,
    "line": 16,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 12,
    "line": 16,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 13,
    "line": 16,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 17,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 17,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 17,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 11,
    "line": 17,
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "column": 12,
    "isBuiltin": true,
    "line": 17,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 20,
    "line": 17,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 21,
    "line": 17,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 22,
    "line": 17,
    "type": "WHITESPACE",
    "value": "

",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 19,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 19,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 19,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 7,
    "line": 19,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 20,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 20,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 20,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 11,
    "line": 20,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 21,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 21,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 21,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 7,
    "line": 21,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 22,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 22,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 22,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 11,
    "line": 22,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 23,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 23,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 23,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 7,
    "line": 23,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 24,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 24,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 24,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 11,
    "line": 24,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 25,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 25,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 25,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 7,
    "line": 25,
    "type": "WHITESPACE",
    "value": "
",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 26,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 26,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 26,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 11,
    "line": 26,
    "type": "WHITESPACE",
    "value": "
",
  },
]
`;

exports[`tokenizer program 'wall' 2`] = `
Array [
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 3,
    "type": "ROUTINE",
    "value": "routine",
  },
  Object {
    "column": 9,
    "isBuiltin": false,
    "line": 3,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 13,
    "line": 3,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 14,
    "line": 3,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 5,
    "isBuiltin": false,
    "line": 4,
    "type": "WHILE",
    "value": "while",
  },
  Object {
    "column": 11,
    "isBuiltin": false,
    "line": 4,
    "type": "NOT",
    "value": "not",
  },
  Object {
    "column": 15,
    "isBuiltin": true,
    "line": 4,
    "type": "IDENTIFIER",
    "value": "isEdge",
  },
  Object {
    "column": 21,
    "line": 4,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 22,
    "line": 4,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 24,
    "isBuiltin": false,
    "line": 4,
    "type": "DO",
    "value": "do",
  },
  Object {
    "column": 9,
    "isBuiltin": true,
    "line": 5,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 19,
    "line": 5,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 20,
    "line": 5,
    "type": "INTEGER",
    "value": "2",
  },
  Object {
    "column": 21,
    "line": 5,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 9,
    "isBuiltin": true,
    "line": 6,
    "type": "IDENTIFIER",
    "value": "step",
  },
  Object {
    "column": 13,
    "line": 6,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 14,
    "line": 6,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 5,
    "line": 7,
    "type": "ASTERISC",
    "value": "*",
  },
  Object {
    "column": 6,
    "isBuiltin": false,
    "line": 7,
    "type": "WHILE",
    "value": "while",
  },
  Object {
    "column": 1,
    "line": 8,
    "type": "ASTERISC",
    "value": "*",
  },
  Object {
    "column": 2,
    "isBuiltin": false,
    "line": 8,
    "type": "ROUTINE",
    "value": "routine",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 11,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 11,
    "line": 11,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 12,
    "line": 11,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 12,
    "type": "IDENTIFIER",
    "value": "step",
  },
  Object {
    "column": 5,
    "line": 12,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 12,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 13,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 11,
    "line": 13,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 12,
    "line": 13,
    "type": "INTEGER",
    "value": "2",
  },
  Object {
    "column": 13,
    "line": 13,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 14,
    "type": "IDENTIFIER",
    "value": "step",
  },
  Object {
    "column": 5,
    "line": 14,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 14,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 15,
    "type": "IDENTIFIER",
    "value": "turnRight",
  },
  Object {
    "column": 10,
    "line": 15,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 11,
    "line": 15,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 13,
    "isBuiltin": true,
    "line": 15,
    "type": "IDENTIFIER",
    "value": "turnRight",
  },
  Object {
    "column": 22,
    "line": 15,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 23,
    "line": 15,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 16,
    "type": "IDENTIFIER",
    "value": "placeBlock",
  },
  Object {
    "column": 11,
    "line": 16,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 12,
    "line": 16,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 17,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 17,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 17,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 12,
    "isBuiltin": true,
    "line": 17,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 20,
    "line": 17,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 21,
    "line": 17,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 19,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 19,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 19,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 20,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 20,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 20,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 21,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 21,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 21,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 22,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 22,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 22,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 23,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 23,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 23,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 24,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 24,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 24,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": false,
    "line": 25,
    "type": "IDENTIFIER",
    "value": "wall",
  },
  Object {
    "column": 5,
    "line": 25,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 6,
    "line": 25,
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "column": 1,
    "isBuiltin": true,
    "line": 26,
    "type": "IDENTIFIER",
    "value": "turnLeft",
  },
  Object {
    "column": 9,
    "line": 26,
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "column": 10,
    "line": 26,
    "type": "RPAREN",
    "value": ")",
  },
]
`;
